### PR TITLE
Add NN model, allow LR model to be trained with same training data, updated tests

### PR DESCRIPTION
**Pull request by lucylw**
* Two NN models have been added (w/ and w/o context), both currently using averaging of list fields like alias and context
* OntoEmma framework and feature generator have been modified to allow LR model to be trained using same training data
* Tests were merged together and unnecessary test files deleted

It's a big one, sorry, but most of the changed files are deleted test files

### CODE REVIEW

File: emma/OntoEmma.py [hunk 1]
```diff
@@ -184,33 +182,49 @@ def load_alignment(self, gold_path):
                 )
 
     def train(
-        self, model_type: str, model_path: str, config_file: str
+        self, model_type: str, model_path: str, config_file: str, cuda_device: int = -1
```
Comment (wammar) [inline emma/OntoEmma.py]: nit: Should `cuda_device` be specified in the config file instead?

File: emma/OntoEmma.py [hunk 2]
```diff
@@ -315,18 +298,132 @@ def train(
 
         return
 
-    def _align_lr(self, model_path, source_kb, target_kb, candidate_selector, feature_generator):
+    def evaluate(self, model_type: str, model_path: str, evaluation_data_file: str, cuda_device: int):
```
Comment (wammar) [inline emma/OntoEmma.py:308]: The `train` method had a default value for `cuda_device`. Should `evaluate` also have it?

File: emma/OntoEmma.py (hunk 1 above)
Comment (wammar) [reply emma/OntoEmma.py]: Just realized that the value specified for `cuda_device` here overrides the configs. Is this necessary? I find it very confusing and likely to lead to unintended behavior.

File: emma/allennlp_classes/ontoemma_model.py [hunk 3]
```diff
@@ -19,46 +20,67 @@ class OntoEmmaNN(Model):
...
-                                     "input dimension of the siamese_feedforward net. Found {} and {}, "
-                                     "respectively.".format(name_rnn_encoder.get_output_dim(),
-                                                            siamese_feedforward.get_input_dim()))
-
         self.accuracy = BooleanF1()
         self.loss = torch.nn.BCELoss()
 
         initializer(self)
 
+    @staticmethod
```
Comment (DeNeutoy) [inline emma/allennlp_classes/ontoemma_model.py:50]: You should do this masked averaging using [this](https://github.com/allenai/allennlp/blob/master/allennlp/modules/seq2vec_encoders/boe_encoder.py#L11), as this way is not very efficient as it isn't vectorised. Should be just `BagOfEmbeddingsEncoder(dim, averaged=True)(tensor, mask)`. [This line](https://github.com/allenai/allennlp/blob/master/allennlp/modules/seq2vec_encoders/boe_encoder.py#L54) means that it will handle zero length sequences correctly, so all your `ListFields` should survive intact.

File: emma/allennlp_classes/ontoemma_model.py [hunk 4]
```diff
@@ -19,46 +20,67 @@ class OntoEmmaNN(Model):
...
-                 initializer: InitializerApplicator = InitializerApplicator()) -> None:
-        super(OntoEmmaNN, self).__init__(vocab)
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+        super(OntoEmmaNN, self).__init__(vocab, regularizer)
 
         self.name_text_field_embedder = name_text_field_embedder
-        self.num_classes = self.vocab.get_vocab_size("labels")
+        self.distributed_name_embedder = BasicTextFieldEmbedder({
+            k: TimeDistributed(v) for k, v in name_text_field_embedder._token_embedders.items()
```
Comment (DeNeutoy) [inline emma/allennlp_classes/ontoemma_model.py:34]: @matt-gardner Do you think this is an abstraction worth adding to `AllenNLP`? It's nice that it just works, but we could add a `time_distribute` flag to `forward` which would mean this was not necessary.

File: config/ontoemma_nn_all.json [hunk 5]
```diff
@@ -0,0 +1,107 @@
+{
```
Comment (DeNeutoy) [inline config/ontoemma_nn_all.json:1]: I've added some config suggestions below - these are only suggestions, feel free to ignore.

File: config/ontoemma_nn_all.json [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            "tokens": {
+                    "type": "embedding",
+                    "pretrained_file": "/net/nfs.corp/s2-research/scigraph/data/ontoemma/embeddings/weights100.txt.gz",
+                    "embedding_dim": 100,
+                    "trainable": false
+            },
+            "token_characters": {
+                    "type": "character_encoding",
+                    "embedding": {
+                      "embedding_dim": 8
```
Comment (DeNeutoy) [inline config/ontoemma_nn_all.json:34]: You might want to consider making this character embedding a little larger - it seems that character based features could be important for drug names, abbreviations etc.

File: config/ontoemma_nn_all.json [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      "dropout": 0.2
+    },
+    "decision_feedforward": {
+      "input_dim": 200,
+      "num_layers": 2,
+      "hidden_dims": [200, 1],
+      "activations": ["relu", "linear"],
+      "dropout": [0.2, 0.0]
+    }
+  },
```
Comment (DeNeutoy) [inline config/ontoemma_nn_all.json:83]: You might want to add an `initializer` field here. Here are a few suggestions:
- Use `xavier_normal` intialisation for all of your feedforward layers and the input -> hidden_state weights of the RNNs.
- Use `orthogonal` initialisation for the hidden_state -> hidden_state weights in the RNNs. These are called something very cryptic in pytorch, like `weight_hh`, so you might have to dig around a little to find out what the names are. If you look in the allennlp logs the parameter names should be logged.

File: config/ontoemma_nn_all.json [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                    },
+                    "dropout": 0.2
+            }
+    },
+    "context_text_field_embedder": {
+            "tokens": {
+                    "type": "embedding",
+                    "pretrained_file": "/net/nfs.corp/s2-research/scigraph/data/ontoemma/embeddings/weights100.txt.gz",
+                    "embedding_dim": 100,
+                    "trainable": false
```
Comment (DeNeutoy) [inline config/ontoemma_nn_all.json:50]: If you aren't training the embeddings, you can probably afford to make the LSTMs a little larger, as you have quite a bit of data (if the bottleneck here is speed, ignore this). Also, did you resolve that CUDNN issue? If not, we should open an issue in Pytorch. Let me know if you want me to dig into that more, particularly if it is the only thing stopping you increasing the batch size to train this thing faster.

**Review by DeNeutoy — COMMENTED** [review]
Just reviewed a few AllenNLP things, sorry for being nosey - it's awesome how you've picked it up and started running (sprinting) with it. Looking forward to seeing how well this works!

File: emma/allennlp_classes/ontoemma_model.py (hunk 4 above)
Comment (matt-gardner) [reply emma/allennlp_classes/ontoemma_model.py:34]: Hmm, I thought that our text field embedder would just work for a `ListField[TextField]`.  Looks like I'm wrong.  Yes, I agree that we should make this unnecessary.

File: config/ontoemma_nn_all.json (hunk 5 above)
Comment (lucylw) [reply config/ontoemma_nn_all.json:1]: Thanks so much for all the suggestions @DeNeutoy !

File: config/ontoemma_nn_all.json (hunk 8 above)
Comment (lucylw) [reply config/ontoemma_nn_all.json:50]: I still can't increase the batch size much :/ I think the way TimeDistributed unfurls the data still causes the cudnn contiguous error, which is unfortunate since I'm only using about half the memory on the GPU. Let me know if you have any thoughts on how I could improve this.

File: emma/allennlp_classes/ontoemma_model.py (hunk 3 above)
Comment (lucylw) [reply emma/allennlp_classes/ontoemma_model.py:50]: Ah that's great; I was wondering whether there was a more efficient existing solution for this. Thanks!

File: emma/FeatureGenerator.py [hunk 9]
```diff
@@ -11,22 +11,18 @@
 
 # class for generating features between entities of two KBs
 class FeatureGenerator:
-    def __init__(self, s_kb: KnowledgeBase, t_kb: KnowledgeBase):
-        self.SYNONYM_REL_LABELS = ['RL', 'RQ', 'RU', 'SY']
-        self.PARENT_REL_LABELS = ['RB', 'PAR', 'Is a', 'Part of', 'subClassOf', 'is_a', 'part_of']
-        self.CHILD_REL_LABELS = ['RN', 'CHD', 'Has part', 'subClass', 'has_part']
-        self.SIBLING_REL_LABELS = ['SIB', 'RO']
+    def __init__(self, data):
```
Comment (wammar) [inline emma/FeatureGenerator.py]: Consider adding a brief description of what `data` should look like.

File: emma/OntoEmma.py [hunk 10]
```diff
@@ -184,33 +182,49 @@ def load_alignment(self, gold_path):
...
         :param model_path: path to ontoemma model
         :param config_file: path to training data, dev data, config for nn
+        :param cuda_device: GPU device number
         :return:
         """
         assert model_type in constants.IMPLEMENTED_MODEL_TYPES
         assert config_file is not None
         assert os.path.exists(config_file)
 
+        # train NN model with AllenNLP
```
Comment (wammar) [inline emma/OntoEmma.py:207]: `OntoEmma.train` is getting a bit too complex. Consider refactoring the following block as follows to improve readability and encapsulate model-specific code.
```
if model_type == 'nn':
  self._train_nn()
elif model_type == 'lr':
  self._train_lr()
else:
  raise NotImplementedError
```

Comment (wammar) [reply emma/OntoEmma.py:207]: nit: consider replacing `'nn'` everywhere with `constants.NN_MODEL_TYPE`. Then you can use a more expressive name and get rid of the `constants.IMPLEMENTED_MODEL_TYPES` map.

File: emma/OntoEmma.py [hunk 11]
```diff
@@ -184,33 +182,49 @@ def load_alignment(self, gold_path):
...
         if model_type == "nn":
             sys.stdout.write("Training {} model...\n".format(constants.IMPLEMENTED_MODEL_TYPES[model_type]))
 
+            from allennlp.commands.train import train_model_from_file
```
Comment (wammar) [inline emma/OntoEmma.py]: nit: can we move this imports at the beginning of the file? (since this is called before you register allennlp ontoemma classes)

File: emma/OntoEmma.py [hunk 12]
```diff
@@ -225,76 +239,47 @@ def train(
...
         # load training data
-        training_data = self.load_alignment(training_data_path)
-        training_pairs = [(ent1, ent2) for ent1, ent2, _ in training_data]
-        training_labels = [label for _, _, label in training_data]
-
-        training_ordered_indices = []
-        training_features = []
+        training_pairs = []
+        training_labels = []
+        with jsonlines.open(training_data_path) as reader:
```
Comment (wammar) [inline emma/OntoEmma.py:251]: Kudos for standardizing the file formats for both models!

File: emma/OntoEmma.py [hunk 13]
```diff
@@ -225,76 +239,47 @@ def train(
...
-            # load kbs if matches not empty
-            if len(training_matches) > constants.MIN_TRAINING_SET_SIZE:
-                training_ordered_indices += training_matches
-                dev_ordered_indices += dev_matches
-                sys.stdout.write(
-                    "\tCalculating features for pairs between %s and %s\n" %
-                    (s_kb_name, t_kb_name)
-                )
+        # initialize feature generator
+        feat_gen = FeatureGenerator([item for sublist in training_pairs for item in sublist])
```
Comment (wammar) [inline emma/OntoEmma.py]: nit: consider renaming `FeatureGenerator` to clarify that it's only used for the LR model. `LRFeatureGenerator` maybe?

File: emma/FeatureGenerator.py [hunk 14]
```diff
@@ -11,22 +11,18 @@
 
 # class for generating features between entities of two KBs
```
Comment (wammar) [inline emma/FeatureGenerator.py]: nit: I think the description needs to be updated after this refactoring.

File: emma/FeatureGenerator.py (hunk 9 above)
Comment (wammar) [reply emma/FeatureGenerator.py]: I don't understand why we need to specify `data` when we instantiate a `FeatureGenerator` object. It'd be a lot easier to use this class if it processes each entity lazily when needed (and cash the features so that we don't recompute things in each epoch), e.g., 
```
featurizer = FeatureGenerator()
for source_entity, target_entity in pairs:
  pair_features = featurizer.generate_features(source_entity, target_entity)
  # use pair_features
```

File: emma/OntoEmma.py [hunk 15]
```diff
@@ -315,18 +298,132 @@ def train(
...
+        :param cuda_device: GPU device number
+        :return:
+        """
+        sys.stdout.write("Begin evaluation...\n")
+
+        assert os.path.exists(model_path)
+        assert evaluation_data_file is not None
+        assert os.path.exists(evaluation_data_file)
+
+        if model_type == "nn":
```
Comment (wammar) [inline emma/OntoEmma.py:323]: Similar to my previous comment in `OntoEmma.train`, consider refactoring this into 
```
if model_type == 'nn':
  self._evaluate_nn()
elif model_type == 'lr':
  self._evaluate_lr()
else:
  raise NotImplementedError
```

File: emma/OntoEmma.py [hunk 16]
```diff
@@ -315,18 +298,132 @@ def train(
...
+            )
+
+        tp = 0
+        fp = 0
+        tn = 0
+        fn = 0
+
+        # iterature through evaluation examples
+        for features, label in zip(eval_features, eval_labels):
+            prediction = model.predict_entity_pair(features)
```
Comment (wammar) [inline emma/OntoEmma.py:385]: I think that using `OntoEmmaNNModel` messed up the integrity/encapsulation of the `OntoEmma` class. It's a shallow wrapper for `OntoEmmaNNModel`, but it has a lot of work that complements `OntoEmmaLRModel` (e.g., evaluation, featurizing data instances). Ideally, we'd move most of the `OntoEmma` code specific to the LR model into `OntoEmmaLRModel`. That said, it's a big refactoring and it has a lower priority than the other things on your plate in the remaining time. Could you add a TODO comment at the beginning of the `OntoEmma` class definition (with the content of this comment) to document this?

File: emma/constants.py [hunk 17]
```diff
@@ -21,5 +21,23 @@
...
+# training, development, and test split
+TRAINING_PART = 0.6
+DEVELOPMENT_PART = 0.2
+TEST_PART = 0.2
+
+# relation labels from UMLS
+UMLS_SYNONYM_REL_LABELS = ['RL', 'RQ', 'RU', 'SY']
+UMLS_PARENT_REL_LABELS = ['RB', 'PAR', 'Is a', 'Part of', 'subClassOf', 'is_a', 'part_of']
+UMLS_CHILD_REL_LABELS = ['RN', 'CHD', 'Has part', 'subClass', 'has_part']
+UMLS_SIBLING_REL_LABELS = ['SIB', 'RO']
```
Comment (wammar) [inline emma/constants.py:43]: It's not clear which constants should be specified in `constants.py` and which should be specified in the config files. If there is a clear distinction, please explain it in the readme file. Otherwise, we might as well move everything to the config files.

**Review by wammar — APPROVED** [review]
Sorry it took me a while to review, LGTM!

File: emma/FeatureGenerator.py (hunk 9 above)
Comment (lucylw) [reply emma/FeatureGenerator.py]: I added a TODO comment here and will refactor later.

File: emma/OntoEmma.py (hunk 1 above)
Comment (lucylw) [reply emma/OntoEmma.py]: I did this initially because it was easier to change cuda devices quickly for experiments, but I think it makes more sense to have the cuda device specified in the config file and nowhere else. I will change this along with some of the other refactoring below.

File: emma/OntoEmma.py (hunk 10 above)
Comment (lucylw) [reply emma/OntoEmma.py:207]: I completely agree. I actually already refactored in this way for train/evaluate/align in another branch, so I will probably open another PR with it instead of trying to apply it to this branch.

File: emma/OntoEmma.py (hunk 11 above)
Comment (lucylw) [reply emma/OntoEmma.py]: done. moved all the allennlp classes to the beginning
